src/kakarot/stack.cairo

```diff
@@ -77,30 +91,110 @@ namespace Stack {
         new_stack: model.Stack*, elements: Uint256*
     ) {
         alloc_locals;
-        Stack.check_underflow(self=self, stack_index=n - 1);
-        // Get new segment for next stack copy
+        let stack_word_dict = self.stack_word_dict;
+        let position_zero = self.stack_16bytes_len;
+
+        // Check if there is underflow
+        with_attr error_message("Kakarot: StackUnderflow") {
+            assert_le(n * 2, position_zero);
+        }
+
         let (new_elements: Uint256*) = alloc();
-        // Get length of new stack copy
-        let new_len = self.raw_len - (element_size * n);
-        // Copy stack without last N elements
-        memcpy(dst=new_elements, src=self.elements, len=new_len);
-
-        // Create new stack
-        local new_stack: model.Stack* = new model.Stack(elements=new_elements, raw_len=new_len);
-        // Return new stack & pointer to first popped element
-        return (new_stack=new_stack, elements=self.elements + new_len);
+        // Read and Copy the elements on an array
+
+        // Generate an array of Uint256* to return
+        let (stack_word_dict) = dict_copy(
+            stack_word_dict=stack_word_dict, stack_len=position_zero, n=n * 2, output=new_elements
+        );
+
+        // Return Stack with updated Len
+        let popped_len = 2 * n;
+        return (
+            new model.Stack(
+            stack_word_dict_start=self.stack_word_dict_start,
+            stack_word_dict=stack_word_dict,
+            stack_16bytes_len=self.stack_16bytes_len - popped_len,
+            ),
+            new_elements,
+        );
+    }
+
+    func dict_copy{range_check_ptr}(
```

Comment:
is this a generic `dict` helper or smth only relevant for the stack context? naming is confusing

Reply:
should add docstring

Reply:
so if I'm right, this function is more like `stack_to_uint256`?

Reply:
Yes, is more specific to Stack as it is considering packs of 128bits